### PR TITLE
feat(client-go): decode json error messages

### LINKED ISSUE #3739: deis login - fix "bad request" message to be user readable
As a user, I would like a clearer message when my username and/or password fail on login.

Today reads as follows:
 "400 BAD REQUEST
{u'non_field_errors': [u'Unable to log in with provided credentials.']}

Recommended:
"Unable to login: 
The username and password you entered do not match"

SEE image: http://cl.ly/image/143z3J2W202o
deis be kind

### PR DESCRIPTION
**Pull request by Joshua-Anderson**
Fixes #3739

### CODE REVIEW

**Comment by carmstrong** [comment]
Does this close #3739?

**Comment by Joshua-Anderson** [comment]
@carmstrong Yes, it does.

**Comment by bacongobbler** [comment]
From the logs so nobody has to grok:

> FAIL	github.com/deis/deis/client-go/controller/client [build failed]

**Comment by Joshua-Anderson** [comment]
I fixed up the error and added additional tests.

File: client-go/controller/client/http.go [hunk 1]
```diff
@@ -81,20 +83,57 @@ func (c Client) Request(method string, path string, body []byte) (*http.Response
...
-		return "", -1, err
+		return "", err
 	}
-	return string(resBody), res.StatusCode, nil
+	return string(resBody), checkForErrors(res, string(resBody))
+}
+
+func checkForErrors(res *http.Response, body string) error {
+	switch res.StatusCode {
+	case http.StatusOK, http.StatusCreated, http.StatusNoContent:
```
Comment (bacongobbler) [inline client-go/controller/client/http.go]: we should probably change this to be anything in the 200 status range

**Comment by Joshua-Anderson** [comment]
Unfortunately, I discovered that some of the errors now printed are too generic to be useful: `Error: This field must be unique.` So I had to add in the key so now errors are printed like this:


```
Error:
email: This filed must be unique
```

File: client-go/controller/client/auth.go [hunk 2]
```diff
@@ -129,16 +129,12 @@ func Passwd(username string, password string, newPassword string) error {
 		return err
 	}
 
-	resBody, status, err := client.BasicRequest("POST", "/v1/auth/passwd/", body)
+	_, err = client.BasicRequest("POST", "/v1/auth/passwd/", body)
```
Comment (mboersma) [inline client-go/controller/client/auth.go]: This could be collapsed with the `if err != nil` statement below, same for lines 150 and 182.

**Comment by mboersma** [comment]
This works well:
```console
$ deis create deis
Creating Application... Error: 
id: deis is a reserved name.
400 BAD REQUEST
$ deis login http://deis.local3.deisapp.com --username=fake --password=fake
Error: 
non_field_errors: Unable to log in with provided credentials.
400 BAD REQUEST
```

Code LGTM, modulo the couple of comments outstanding.

**Comment by Joshua-Anderson** [comment]
I compacted error checks (they were all over the code, as many methods now only check for errors) and cleaned up a flakey tests (iteration over the json map is random).

**Comment by bacongobbler** [comment]
code LGTM

File: client-go/controller/client/http.go (hunk 1 above)
Comment (mboersma) [inline client-go/controller/client/http.go]: Should this still be changed so we're not just catching a few specific HTTP codes?

Comment (bacongobbler) [inline client-go/controller/client/http.go]: I think so. Anything the API throws at us in the future should work. Checking the range of non-error response codes would be an easy enhancement.

**Pull request merged**